### PR TITLE
* add SIGTERM handler to gracefully terminate task executors, \

### PR DESCRIPTION
**Pull request by aykut-bozkurt**
DESCRIPTION: Adds signal handlers for queue monitor to gracefully shutdown, cancel and to see config changes.

That PR adds signal handlers for graceful termination, cancellation of task executors and detecting config updates. Related to PR #6459.

#### How to handle termination signal?
Monitor need to gracefully terminate all running task executors before terminating. Hence, we have sigterm handler for the monitor.

#### How to handle cancellation signal?
Monitor need to gracefully cancel all running task executors before terminating. Hence, we have sigint handler for the monitor.

#### How to detect configuration changes?
Monitor has SIGHUP handler to reflect configuration changes while executing tasks.

### CODE REVIEW

File: src/backend/distributed/utils/background_jobs.c [hunk 1]
```diff
@@ -674,22 +735,49 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
...
+			ereport(LOG, (errmsg("handling termination signal")));
+			TerminateAllTaskExecutors(backgroundExecutorHandles);
+			termHandleStarted = true;
+		}
+
+		if (GotSighup)
+		{
+			/* update max_background_task_executors if changed */
+			ProcessConfigFile(PGC_SIGHUP);
+			GotSighup = false;
```
Comment (CyberDem0n) [inline src/backend/distributed/utils/background_jobs.c:766]: For me it looks like there is small race condition.
E.g., while the `ProcessConfigFile()` is running the second SIGHUP may arrive and it would be ignored.

File: src/backend/distributed/utils/background_jobs.c [hunk 2]
```diff
@@ -741,6 +747,53 @@ TaskEnded(TaskExecutionContext *taskExecutionContext)
 }
 
 
+/*
+ * QueueMonitorSigHupHandler handles SIGHUP to update monitor related config params.
+ */
+static void
+QueueMonitorSigHupHandler(SIGNAL_ARGS)
+{
+	GotSighup = true;
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:756]: The signal handlers for the normal maintenance daemon contain some code related to saving `errno`. That sounds like it could be important. Probably nice to share these handlers.

File: src/backend/distributed/utils/background_jobs.c [hunk 3]
```diff
@@ -758,15 +811,25 @@ TaskEnded(TaskExecutionContext *taskExecutionContext)
 void
 CitusBackgroundTaskQueueMonitorMain(Datum arg)
 {
+	/* handle SIGTERM to properly kill active task executors */
+	pqsignal(SIGTERM, QueueMonitorSigTermHandler);
+
+	/* ignore SIGINT signal */
+	pqsignal(SIGINT, SIG_IGN);
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:818]: Why is this necessary? What happens otherwise? This is not done for the maintenance daemon. Should we do it there too?

File: src/backend/distributed/utils/background_jobs.c [hunk 4]
```diff
@@ -946,6 +1033,11 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
 	MemoryContextSwitchTo(firstContext);
 	MemoryContextDelete(backgroundTaskContext);
 
+	if (GotSigterm)
+	{
+		proc_exit(1);
+	}
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:1039]: I think it's common for correctly handled SIGTERM to return a 0 exit code, so I don't think this code is necessary.

File: src/test/regress/expected/background_task_queue_monitor.out [hunk 5]
```diff
@@ -302,6 +308,195 @@ SELECT job_id, task_id, status FROM pg_dist_background_task
...
+SELECT task_id, status, retry_count, message FROM pg_dist_background_task
+    WHERE task_id IN (:task_id1, :task_id2)
+    ORDER BY task_id; -- show that all tasks are runnable by retry policy after termination signal
+ task_id |  status  | retry_count |                                                            message
+---------------------------------------------------------------------
+      21 | runnable |           1 | FATAL: terminating background worker "Citus Background Task Queue Executor: regression/postgres" due to administrator command+
+         |          |             | CONTEXT: Citus Background Task Queue Executor: regression/postgres                                                           +
+         |          |             |
+      22 | runnable |           1 | FATAL: terminating background worker "Citus Background Task Queue Executor: regression/postgres" due to administrator command+
+         |          |             | CONTEXT: Citus Background Task Queue Executor: regression/postgres                                                           +
```
Comment (JelteF) [inline src/test/regress/expected/background_task_queue_monitor.out:463]: sidenote, Now that we can have multiple concurrent background task executors, I think the background the worker names should include the task_id. This can be useful for debugging.

File: src/backend/distributed/utils/background_jobs.c (hunk 3 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/utils/background_jobs.c:818]: SIGINT is sent by pg_backend_cancel. But, we donot want to cancel queue monitor upon recieving the signal.
Default handler for bgworker which runs a backend is like below:

```
if (worker->bgw_flags & BGWORKER_BACKEND_DATABASE_CONNECTION)
{
       /*
	* SIGINT is used to signal canceling the current action
	*/
       pqsignal(SIGINT, StatementCancelHandler);
```

File: src/backend/distributed/utils/background_jobs.c (hunk 2 above)
Comment (JelteF) [reply src/backend/distributed/utils/background_jobs.c:761]: Maybe we can really make these functions shared across the different background daemons that we have? Because they have the exact same body afaict.

File: src/backend/distributed/utils/background_jobs.c [hunk 6]
```diff
@@ -861,22 +962,50 @@ CitusBackgroundTaskQueueMonitorMain(Datum arg)
...
+		 * runs of the same task on the next start of the monitor, which is dangerous for non-idempotent
+		 * tasks. We do not break the loop here as we want to reflect tasks' messages. Hence, we wait until
+		 * all tasks finish and also do not allow new runnable tasks to start running. After all current tasks
+		 * finish, we can exit the loop safely.
+		 */
+		if (MonitorGotTerminationOrCancellationRequest() && !terminateExecutorsStarted)
+		{
+			ereport(LOG, (errmsg("handling termination signal")));
+			terminateExecutorsStarted = true;
+			TerminateAllTaskExecutors(queueMonitorExecutionContext.currentExecutors);
```
Comment (JelteF) [inline src/backend/distributed/utils/background_jobs.c:986]: I think if we get a SIGINT we should forward the SIGINT to the executors.

File: src/backend/distributed/utils/background_jobs.c (hunk 2 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/utils/background_jobs.c:761]: I think I should investigate that in another PR.
